content/en/blog/fsm2/index.md

```diff
@@ -1,207 +1,357 @@
+As the sequence diagram above shows, the service FSM conceptually presents a
+chat room and works as a mediator between chat room participants.
+
+The next Go code block shows how the above FSM instances are declared at the
+programming level. Note that `ServiceFSM` is optional and seldom needed. The
+rest of the data is pretty obvious: 
+- gRPC connection, `conn`
+- conversation FSM, `md`
+- and interrupt channel is given as a startup argument, `intCh`, because here,
+the machine is started from the CLI tool
+
+```go 
+	intCh := make(chan os.Signal, 1)
+	signal.Notify(intCh, syscall.SIGTERM, syscall.SIGINT)
+
+	chat.Bot{
+		Conn:        conn,      // gRPC connection including JWT, etc.
+		MachineData: md,        // primary FSM data, i.e. pairwise lvl
+		ServiceFSM:  mdService, // optional backend/service lvl FSM
+	}.Run(intCh)                // let machine to know when it's time to quit
 ```
 
-For other agent use two terminals and give these commands to them:
-```console
-# terminal 1
-findy-agent-cli bot read # listens and show other end's messages
-# terminal 2
-findy-agent-cli bot chat # sends basic_message's to other end thru the pairwise
-```
 
-And when you want to render your state machine in UML, give this command:
-```console
-findy-agent-cli bot uml <Hello-World.yaml> # or name of your FSM
+By utilizing the CSP model, we embrace a paradigm that emphasizes the
+coordination and communication between independently executing components, known
+as goroutines, in a safe and efficient manner. This combination of Go's
+built-in concurrency primitives, empowers us developers to create
+highly concurrent systems while maintaining clarity and simplicity in their
+code.
+
+## CSP & Go Code
+
+The app logic is in the state machines written in YAML and Lua. But
+surprising is how elegant the Go implementation can be when CSP is
+used for the state-machine processor. And all of that without a single mutex or
+other synchronization object.
+
+The code block below is the crucial component of the FSM engine solution. It runs in
+its goroutine and serves all the input and output at the process level. For
+those who come from traditional multi-thread programming, this might look weird.
+You might ask why a lock doesn't make conversations map thread-safe.
+
+That's the beauty of the CSP. Only this goroutine modifies `conversations`
+data—no one else. You might ask if there's a performance penalty in this specific
+solution, but there is not. The `Multiplexer` function doesn't do anything
+that's computationally extensive. It listens to several Go channels and
+delegates the work to other goroutines.
+
+This model has proven to be easy to understand and implement.
+
+```go
+func Multiplexer(info MultiplexerInfo) {
+	glog.V(3).Infoln("starting multiplexer", info.ConversationMachine.FType)
+	termChan := make(fsm.TerminateChan, 1)
+
+	var backendChan fsm.BackendInChan
+	if info.BackendMachine.IsValid() {
```

Comment:
Is this block relevant for the point you are making? I mean that the code block is quite long and if "irrelevant parts" could be clipped of with "..." it might be clearer.